src/tactic/protect_proj.lean

```diff
@@ -0,0 +1,34 @@
+/-
+Copyright (c) 2020 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+-/
+import meta.expr
+/-!
+# protect_proj user attribute
+
+Attribute to protect the projections of a structure.
+If a structure `foo` is marked with the `protect_proj` user attribute, then
+all of the projections become projected, meaning they must always be referred to by
+their full name `foo.bar`, even when the `foo` namespace is open.
+-/
+
+namespace tactic
+
+/-- Tactic that is executed when a structure is marked with the `protect_proj` attribute -/
+meta def protect_proj_tac (n : name) (env : environment) : tactic environment :=
```

Comment:
I feel like I've noticed this before, do we still not have `tactic.mk_protected`? This seems like a big oversight.

I'd add this function to `tactic.core` and implement `protect_proj_tac : name -> tactic unit` that fails (with your current error message) if the name is not a structure. (There's `environment.is_structure` to test this.)